pkg/executor/bacalhau/client.go

```diff
@@ -1,73 +1,269 @@
+func (c *BacalhauClient) GetVersion() (string, error) {
+	response, err := c.api.Agent().Version(context.Background())
+	if err != nil {
+		return "", fmt.Errorf("error getting bacalhau version %s", err.Error())
+	}
+
+	return response.BuildVersionInfo.GitVersion, nil
+}
+
+func (c *BacalhauClient) PostJob(job bacalhau.Job) (*apimodels.PutJobResponse, error) {
+	
+	translatedJob := translateJob(job)
+	
+	putJobRequest := apimodels.PutJobRequest{
+		Job: translatedJob,
+	}
+	
+	return c.api.Jobs().Put(context.Background(), &putJobRequest)
 }
 
-func getRequest[ResultType any](client *BacalhauClient, requestPath string) (ResultType, error) {
-	var result ResultType
-	url := fmt.Sprintf("%s/%s", client.getBaseUrl(), requestPath)
-	request, err := retryablehttp.NewRequest("GET", url, nil)
+func (c *BacalhauClient) GetJob(jobID string) (*apimodels.GetJobResponse, error) {
+	getJobRequest := apimodels.GetJobRequest{
+		JobID: jobID,
+		Include: "executions",
+	}
+	
+	response, err := c.api.Jobs().Get(context.Background(), &getJobRequest)
 	if err != nil {
-		return result, err
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func (c *BacalhauClient) GetJobResult(jobId string ) (string, error) {
+	getJobResultsRequest := apimodels.ListJobResultsRequest{
+		JobID: jobId,
 	}
-	request.Header.Set("Content-Type", "application/json")
 
-	response, err := client.http.Do(request)
+	response, err := c.api.Jobs().Results(context.Background(), &getJobResultsRequest)
 	if err != nil {
-		return result, err
+		return "", err
+	}
+	if len(response.Items) == 0 {
+		return "", fmt.Errorf("no results found for job %s", jobId)
 	}
-	defer response.Body.Close()
-	body, err := io.ReadAll(response.Body)
+
+	return response.Items[0].Params["URL"].(string), nil
+}
+
+
+
+func (c *BacalhauClient) GetNodes() ([]*models.NodeState, error) {
+	
+	getNodesRequest := apimodels.ListNodesRequest{}
+	response, err := c.api.Nodes().List(context.Background(), &getNodesRequest)
 	if err != nil {
-		return result, err
+		return nil, err
 	}
+	return response.Nodes, nil
+}
 
-	err = json.Unmarshal(body, &result)
+func (c *BacalhauClient) GetMachineSpecs() ([]data.MachineSpec, error) {
+	nodes, err := c.GetNodes()
+	var specs []data.MachineSpec
+	for _, node := range nodes {
+		spec := data.MachineSpec{
+			CPU:  int(node.Info.ComputeNodeInfo.MaxCapacity.CPU) * 1000, // convert float to "milli-CPU"
+			RAM:  int(node.Info.ComputeNodeInfo.MaxCapacity.Memory),
+			GPU:  int(node.Info.ComputeNodeInfo.MaxCapacity.GPU),
+			Disk: int(node.Info.ComputeNodeInfo.MaxCapacity.Disk),
+		}
+		for _, gpu := range node.Info.ComputeNodeInfo.MaxCapacity.GPUs {
+			spec.GPUs = append(spec.GPUs, data.GPUSpec{
+				Name:   gpu.Name,
+				Vendor: string(gpu.Vendor),
+				VRAM:   int(gpu.Memory),
+			})
+		}
+		specs = append(specs, spec)
+	}
 	if err != nil {
-		return result, err
+		return nil, err
 	}
+	return specs, nil
+}
+
+
 
-	return result, nil
+func translateJob(job bacalhau.Job) *models.Job {
```

Comment:
Hurray! So happy to see this start to exist. Does it make sense to have an interface for this for other executors? Or do you see it always being handled internally like this?

Reply:
Leaving this note here for history: translateJob will be deprecated after the new job spec update.